backend/data_preparation/crawler/fire_crawler.py

```diff
@@ -0,0 +1,318 @@
+        """
+        Extracts all fires on the website and return.
+        The returned value is a list of tuples:(year:int, state:string, fire's name in its url)
+        :param start_year: int, start year for the crawler. Record before the start year will be deleted.
+                e.g. 2015
+        :return: list of FireEvent objects
+                e.g. [FireEvent(-1, 2015, 'California', 'FireA'), FireEvent(-1, 2015, 'California', 'FireB')]
+        """
+        logger.info("Attempting to get all links...")
+        current_year = datetime.datetime.now().date().year
+        # fetch all year nodes from the website, may encounter errors
+        # add a try-except block, and while loop to keep trying
+        # if attempts > 3 raise an exception
+        try:
+            main_page = self._get_url(FireCrawler.BASE_DIR, "main_page", "extract_all_fires()")
+        except requests.exceptions.RequestException:
+            logger.error("Error: Cannot fetch main_page in function extract_all_fires()")
+            raise CannotCrawlException("Error: Cannot fetch main_page in function extract_all_fires")
+            # retry exception
+        else:
+            # the last node is "historical_fire_data" and we don't want it
+            sub_links_of_each_year = FireCrawler.RE_EXTRACT_ALL_FIRES.findall(main_page)[:-1]
+            # sub_links_of_each_year are now fire data of a certain year or current year
+            # e.g. '2010_fire_data', 'current_year_fire_data'
+            fires = []
+            for sub_link_of_each_year in sub_links_of_each_year:
+                # sub_link_of_each_year: e.g. '2010_fire_data'
+                # to change "current_year" to the real year as an integer
+                year_as_int = current_year if sub_link_of_each_year == "current_year_fire_data" \
+                                         else int(sub_link_of_each_year.split("_")[0])
+                # ignore years that before the start year
+                # filter out links that is too old
+                if year_as_int >= start_year:
+                    for state in self.states:
+                        # for each state, extract fire events in that year at that state
+                        fires.extend(self._extract_fire_events_in_state(year_as_int, state))
+            self._filter_out_invalid_fire_events(fires)
+            return fires
+
+    @staticmethod
+    def _download_fire_record(single_file_name: str, used_folder_names: Set[str], url_of_fire_event: str):
+        """
+        Downloads a single fire record from the website and puts its sub-files into a folder.
+        :param single_file_name: e.g. "ca_cedar_20170810_0000_dd83.CPG"
+        :param used_folder_names:e.g. ("ca_cedar_20170810_0000_dd83", "ca_cedar_20170812_2145_dd83")
+        :param url_of_fire_event:e.g "https://rmgsc.cr.usgs.gov/outgoing/GeoMAC/2017_fire_data/California/Cedar/"
+        """
+        if not ".zip" == single_file_name[-4:]:
```

Comment:
can we put some log in this function? especially when the if statement is false, and the method is doing nothing, the outer-scope might expect something to be downloaded

Reply:
The helper function called will log, so I put a log on the situation that crawler meets zip files.